File: impl-large/cpp/service.cpp
// Task Service, large tier — every rule that decides, mutates and audits.

#include "service.hpp"

#include <cmath>
#include <set>

std::string serializeUser(const User &user)
{
    return std::format(R"({{"id":{},"username":{},"role":{},"quota":{},"version":{},)"
                       R"("deleted":{}}})",
                       user.id, writeJsonString(user.username), writeJsonString(user.role),
                       user.quota, user.version, boolText(user.deleted));
}

std::string serializeProject(const Project &project)
{
    return std::format(R"({{"id":{},"name":{},"ownerId":{},"taskCount":{},"version":{},)"
                       R"("deleted":{}}})",
                       project.id, writeJsonString(project.name), project.ownerId,
                       taskCount(project.id), project.version, boolText(project.deleted));
}

std::string serializeTask(const Task &task, const std::string &role)
{
    const std::string note =
        role == "admin" ? std::format(R"("internalNote":{},)", writeJsonString(task.internalNote))
                        : std::string{};
    return std::format(R"({{"id":{},"projectId":{},"title":{},"priority":{},"status":{},)"
                       R"("assigneeId":{},{}"version":{},"deleted":{},"score":{}}})",
                       task.id, task.projectId, writeJsonString(task.title), task.priority,
                       writeJsonString(task.status),
                       task.assigneeId ? std::to_string(*task.assigneeId) : std::string{"null"},
                       note, task.version, boolText(task.deleted),
                       computeScore(task.priority, task.status));
}

std::string serializeComment(const Comment &comment)
{
    return std::format(R"({{"id":{},"taskId":{},"authorId":{},"body":{}}})", comment.id,
                       comment.taskId, comment.authorId, writeJsonString(comment.body));
}

std::string serializeAudit(const AuditEntry &entry)
{
    return std::format(R"({{"seq":{},"actorId":{},"action":{},"resource":{},"resourceId":{}}})",
                       entry.seq, entry.actorId, writeJsonString(entry.action),
                       writeJsonString(entry.resource), entry.resourceId);
}

std::string serializeOutbox(const OutboxEvent &event)
{
    return std::format(R"({{"seq":{},"name":{},"resourceId":{},"delivered":{}}})", event.seq,
                       writeJsonString(event.name), event.resourceId, boolText(event.delivered));
}

Caller authenticate(const std::string &header)
{
    if (!header.starts_with("Bearer ")) throw unauthorized();
    const auto found = sessions.find(header.substr(7));
    if (found == sessions.end()) throw unauthorized();
    User *user = findUser(found->second.userId);
    if (user == nullptr) throw unauthorized();
    return {user, &found->second};
}

int chargeQuota(User &user, Session &session)
{
    if (session.used >= user.quota) throw quotaExceeded();
    session.used += 1;
    return std::max(user.quota - session.used, 0);
}

void requireAdmin(const User &user)
{
    if (user.role != "admin") throw forbidden();
}

Project &reachableProject(int projectId, const User &user, bool includeDeleted)
{
    Project *project = findProject(projectId, includeDeleted);
    if (project == nullptr) throw notFound();
    if (user.role != "admin" && project->ownerId != user.id) throw forbidden();
    return *project;
}

Task &reachableTask(int taskId, const User &user, bool includeDeleted)
{
    Task *task = findTask(taskId, includeDeleted);
    if (task == nullptr) throw notFound();
    reachableProject(task->projectId, user, true);
    return *task;
}

void checkIfMatch(const std::optional<std::string> &value, int version)
{
    if (!value || value->empty()) throw preconditionRequired();
    if (*value != std::to_string(version)) throw preconditionFailed();
}

bool checkIncludeDeleted(const std::optional<std::string> &raw, const User &user)
{
    if (!raw) return false;
    if (user.role != "admin") throw forbidden();
    return *raw == "true";
}

bool lessUser(const User &a, const User &b, const std::string &sort)
{
    if (sort == "username") return a.username < b.username;
    if (sort == "role") return a.role < b.role;
    return a.id < b.id;
}

bool lessProject(const Project &a, const Project &b, const std::string &sort)
{
    if (sort == "name") return a.name < b.name;
    if (sort == "taskCount") return taskCount(a.id) < taskCount(b.id);
    return a.id < b.id;
}

bool lessTask(const Task &a, const Task &b, const std::string &sort)
{
    if (sort == "title") return a.title < b.title;
    if (sort == "priority") return a.priority < b.priority;
    if (sort == "score") {
        return computeScore(a.priority, a.status) < computeScore(b.priority, b.status);
    }
    if (sort == "status") return a.status < b.status;
    return a.id < b.id;
}

bool lessComment(const Comment &a, const Comment &b, const std::string &sort)
{
    if (sort == "authorId") return a.authorId < b.authorId;
    return a.id < b.id;
}

bool lessAudit(const AuditEntry &a, const AuditEntry &b, const std::string &)
{
    return a.seq < b.seq;
}

bool lessOutbox(const OutboxEvent &a, const OutboxEvent &b, const std::string &)
{
    return a.seq < b.seq;
}

const User &login(const std::string &username, const std::string &password,
                  const std::string &token)
{
    const User *user = findByUsername(username);
    if (user == nullptr || user->password != password) throw invalidCredentials();
    sessions.emplace(token, Session{token, user->id, 0});
    return *user;
}

Project &createProject(const User &actor, const std::string &name, std::optional<int> ownerId)
{
    Details errors;
    checkString(name, "name", MAX_NAME_LENGTH, errors);
    if (!ownerId || findUser(*ownerId) == nullptr) {
        errors.push_back(fail("ownerId", "ownerId is not a known user"));
    }
    if (!errors.empty()) throw invalid(std::move(errors));
    for (const auto &[id, other] : projects) {
        if (other.ownerId == *ownerId && other.name == name && !other.deleted) throw conflict();
    }
    Project &project = insertProject(name, *ownerId);
    record(actor.id, "create", "project", project.id);
    return project;
}

void renameProject(const User &actor, Project &project, const std::string &name)
{
    Details errors;
    checkString(name, "name", MAX_NAME_LENGTH, errors);
    if (!errors.empty()) throw invalid(std::move(errors));
    for (const auto &[id, other] : projects) {
        if (other.ownerId == project.ownerId && other.name == name && other.id != project.id
            && !other.deleted) {
            throw conflict();
        }
    }
    project.name = name;
    project.version += 1;
    record(actor.id, "update", "project", project.id);
}

void deleteProject(const User &actor, Project &project)
{
    project.deleted = true;
    project.version += 1;
    record(actor.id, "delete", "project", project.id);
    for (Task *task : liveTasksOf(project.id)) {
        task->deleted = true;
        task->version += 1;
        record(actor.id, "delete", "task", task->id);
    }
}

void restoreProject(const User &actor, Project &project)
{
    if (!project.deleted) throw conflict();
    project.deleted = false;
    project.version += 1;
    record(actor.id, "restore", "project", project.id);
}

std::string readNote(const User &actor, const Json &body, Details &errors,
                     const std::string &current)
{
    const Json *note = body.find("internalNote");
    if (note == nullptr) return current;
    if (actor.role != "admin") throw forbidden();
    if (note->kind != Json::Kind::String) throw badRequest();
    if (note->text.size() > MAX_TITLE_LENGTH) {
        errors.push_back(fail("internalNote", "internalNote is too long"));
    }
    return note->text;
}

namespace {

void checkTaskFields(const std::string &title, std::optional<int> priority,
                     std::optional<int> assigneeId, Details &errors)
{
    checkString(title, "title", MAX_TITLE_LENGTH, errors);
    checkPriority(priority, errors);
    if (assigneeId && findUser(*assigneeId) == nullptr) {
        errors.push_back(fail("assigneeId", "assigneeId is not a known user"));
    }
    if (!errors.empty()) throw invalid(std::move(errors));
}

}  // namespace

Task &createTask(const User &actor, const Project &project, const std::string &title,
                 std::optional<int> priority, std::optional<int> assigneeId,
                 const std::string &note, Details &errors)
{
    checkTaskFields(title, priority, assigneeId, errors);
    Task &task = insertTask(project.id, title, *priority, assigneeId, note);
    record(actor.id, "create", "task", task.id);
    return task;
}

void replaceTask(const User &actor, Task &task, const std::string &title,
                 std::optional<int> priority, std::optional<int> assigneeId,
                 const std::string &note, Details &errors)
{
    checkTaskFields(title, priority, assigneeId, errors);
    task.title = title;
    task.priority = *priority;
    task.assigneeId = assigneeId;
    task.internalNote = note;
    task.version += 1;
    record(actor.id, "update", "task", task.id);
}

void moveStatus(const User &actor, Task &task, const Json *status)
{
    Details errors;
    checkStatus(status, errors);
    if (!errors.empty()) throw invalid(std::move(errors));
    if (!allowedMove(task.status, status->text)) throw invalidTransition();
    task.status = status->text;
    task.version += 1;
    record(actor.id, "update", "task", task.id);
}

void deleteTask(const User &actor, Task &task)
{
    task.deleted = true;
    task.version += 1;
    record(actor.id, "delete", "task", task.id);
}

void restoreTask(const User &actor, Task &task)
{
    if (!task.deleted) throw conflict();
    task.deleted = false;
    task.version += 1;
    record(actor.id, "restore", "task", task.id);
}

Comment &createComment(const User &actor, const Task &task, const std::string &body)
{
    Details errors;
    checkString(body, "body", MAX_COMMENT_LENGTH, errors);
    if (!errors.empty()) throw invalid(std::move(errors));
    Comment &comment = insertComment(task.id, actor.id, body);
    record(actor.id, "create", "comment", comment.id);
    return comment;
}

void removeComment(const User &actor, const Comment &comment)
{
    if (actor.role != "admin" && comment.authorId != actor.id) throw forbidden();
    const int commentId = comment.id;
    comments.erase(commentId);
    record(actor.id, "delete", "comment", commentId);
}

User &createUser(const User &actor, const std::string &username, const std::string &password,
                 const Json *role, const Json *quota)
{
    Details errors;
    checkString(username, "username", MAX_NAME_LENGTH, errors);
    checkString(password, "password", MAX_NAME_LENGTH, errors);
    checkRole(role, errors);
    checkQuota(quota, errors);
    if (!errors.empty()) throw invalid(std::move(errors));
    if (findByUsername(username) != nullptr) throw conflict();
    User &user = insertUser(username, password, role == nullptr ? "user" : role->text,
                            quota == nullptr ? DEFAULT_QUOTA : static_cast<int>(quota->number));
    record(actor.id, "create", "user", user.id);
    return user;
}

void updateUser(const User &actor, User &user, const Json &body)
{
    Details errors;
    const Json *role = body.find("role");
    const Json *quota = body.find("quota");
    checkRole(role, errors);
    checkQuota(quota, errors);
    if (!errors.empty()) throw invalid(std::move(errors));
    if (role != nullptr) user.role = role->text;
    if (quota != nullptr) user.quota = static_cast<int>(quota->number);
    user.version += 1;
    record(actor.id, "update", "user", user.id);
}

void deleteUser(const User &actor, User &user)
{
    if (user.id == actor.id) throw conflict();
    user.deleted = true;
    user.version += 1;
    record(actor.id, "delete", "user", user.id);
}

std::vector<const Project *> visibleProjects(const User &user, bool includeDeleted)
{
    std::vector<const Project *> rows;
    for (const auto &[id, project] : projects) {
        if ((includeDeleted || !project.deleted)
            && (user.role == "admin" || project.ownerId == user.id)) {
            rows.push_back(&project);
        }
    }
    return rows;
}

std::vector<const Task *> visibleTasks(const User &user, bool includeDeleted)
{
    std::set<int> allowed;
    for (const Project *project : visibleProjects(user, true)) allowed.insert(project->id);
    std::vector<const Task *> rows;
    for (const auto &[id, task] : tasks) {
        if (allowed.contains(task.projectId) && (includeDeleted || !task.deleted)) {
            rows.push_back(&task);
        }
    }
    return rows;
}

std::string search(const User &user, const std::string &query)
{
    const std::string needle = lower(query);
    std::string results;
    int total = 0;
    const auto emit = [&](std::string_view type, int id, const std::string &label) {
        if (!results.empty()) results += ',';
        results += std::format(R"({{"type":"{}","id":{},"label":{}}})", type, id,
                               writeJsonString(label));
        total += 1;
    };
    for (const Project *project : visibleProjects(user, false)) {
        if (lower(project->name).find(needle) != std::string::npos) {
            emit("project", project->id, project->name);
        }
    }
    for (const Task *task : visibleTasks(user, false)) {
        if (lower(task->title).find(needle) != std::string::npos) {
            emit("task", task->id, task->title);
        }
    }
    return std::format(R"({{"results":[{}],"total":{}}})", results, total);
}

std::string workload(const User &user, const std::string &groupBy)
{
    const std::vector<const Task *> rows = visibleTasks(user, false);
    std::string groups;
    const auto emit = [&groups](std::string_view key, const std::vector<const Task *> &picked) {
        int totalScore = 0;
        for (const Task *task : picked) totalScore += computeScore(task->priority, task->status);
        if (!groups.empty()) groups += ',';
        groups += std::format(R"({{"key":{},"tasks":{},"totalScore":{}}})", writeJsonString(key),
                              picked.size(), totalScore);
    };
    const auto pick = [&rows](auto keep) {
        std::vector<const Task *> picked;
        for (const Task *task : rows) {
            if (keep(task)) picked.push_back(task);
        }
        return picked;
    };
    if (groupBy == "status") {
        for (const auto &entry : STATUS_BONUS) {
            emit(entry.name, pick([&entry](const Task *task) { return task->status == entry.name; }));
        }
    } else if (groupBy == "assignee") {
        std::set<int> named;
        for (const Task *task : rows) {
            if (task->assigneeId) named.insert(*task->assigneeId);
        }
        for (const int assignee : named) {
            emit(std::to_string(assignee),
                 pick([assignee](const Task *task) { return task->assigneeId == assignee; }));
        }
        const std::vector<const Task *> loose =
            pick([](const Task *task) { return !task->assigneeId; });
        if (!loose.empty()) emit("unassigned", loose);
    } else {
        for (const Project *project : visibleProjects(user, false)) {
            emit(project->name,
                 pick([project](const Task *task) { return task->projectId == project->id; }));
        }
    }
    return std::format(R"({{"groupBy":{},"groups":[{}]}})", writeJsonString(groupBy), groups);
}

int flushOutbox()
{
    int flushed = 0;
    for (OutboxEvent &event : outbox) {
        if (!event.delivered) {
            event.delivered = true;
            flushed += 1;
        }
    }
    return flushed;
}

std::string metrics()
{
    std::string statusPart;
    for (const auto &[code, count] : byStatus) {
        if (!statusPart.empty()) statusPart += ',';
        statusPart += std::format(R"("{}":{})", code, count);
    }
    std::string routePart;
    for (const auto &[route, count] : byRoute) {
        if (!routePart.empty()) routePart += ',';
        routePart += std::format(R"({{"route":{},"count":{}}})", writeJsonString(route), count);
    }
    return std::format(R"({{"requests":{},"byStatus":{{{}}},"byRoute":[{}],"auditEntries":{},)"
                       R"("outboxPending":{}}})",
                       requests, statusPart, routePart, audit.size(), outboxPending());
}

std::string stats()
{
    std::string statusPart;
    for (const auto &entry : STATUS_BONUS) {
        int count = 0;
        for (const auto &[id, task] : tasks) {
            if (!task.deleted && task.status == entry.name) count += 1;
        }
        if (!statusPart.empty()) statusPart += ',';
        statusPart += std::format(R"("{}":{})", entry.name, count);
    }
    int total = 0;
    int sumScore = 0;
    for (const auto &[id, task] : tasks) {
        if (task.deleted) continue;
        total += 1;
        sumScore += computeScore(task.priority, task.status);
    }
    const double avgScore =
        total == 0 ? 0.0
                   : std::round(static_cast<double>(sumScore) / static_cast<double>(total) * 100.0)
                         / 100.0;
    const Project *best = nullptr;
    for (const auto &[id, project] : projects) {
        if (project.deleted) continue;
        if (best == nullptr || taskCount(project.id) > taskCount(best->id)) best = &project;
    }
    return std::format(R"({{"projects":{},"tasks":{},"users":{},"sessions":{},"comments":{},)"
                       R"("byStatus":{{{}}},"avgScore":{},"topProjectName":{},)"
                       R"("auditEntries":{},"outboxPending":{}}})",
                       liveProjects(), total, liveUsers(), sessions.size(), comments.size(),
                       statusPart, avgScore,
                       best == nullptr ? std::string{"null"} : writeJsonString(best->name),
                       audit.size(), outboxPending());
}

void checkBulkSize(const Json *operations)
{
    if (operations == nullptr || operations->kind != Json::Kind::Array
        || operations->items.empty() || operations->items.size() > MAX_BULK_ITEMS) {
        throw invalid({fail("operations", "operations is out of range")});
    }
}

File: impl-large/kotlin/src/main/kotlin/com/example/taskservice/Service.kt
// Task Service, large tier — business rules, authorization and audit emission.
package com.example.taskservice

// --------------------------------------------------------------------- serializers

fun serializeUser(user: User): Map<String, Any?> = mapOf(
    "id" to user.id,
    "username" to user.username,
    "role" to user.role,
    "quota" to user.quota,
    "version" to user.version,
    "deleted" to user.deleted,
)

fun serializeTask(task: Task, role: String): Map<String, Any?> = buildMap {
    put("id", task.id)
    put("projectId", task.projectId)
    put("title", task.title)
    put("priority", task.priority)
    put("status", task.status)
    put("assigneeId", task.assigneeId)
    if (role == "admin") put("internalNote", task.internalNote)
    put("version", task.version)
    put("deleted", task.deleted)
    put("score", computeScore(task.priority, task.status))
}

fun serializeComment(comment: Comment): Map<String, Any?> = mapOf(
    "id" to comment.id,
    "taskId" to comment.taskId,
    "authorId" to comment.authorId,
    "body" to comment.body,
)

fun serializeAudit(entry: AuditEntry): Map<String, Any?> = mapOf(
    "seq" to entry.seq,
    "actorId" to entry.actorId,
    "action" to entry.action,
    "resource" to entry.resource,
    "resourceId" to entry.resourceId,
)

fun serializeOutbox(event: OutboxEvent): Map<String, Any?> = mapOf(
    "seq" to event.seq,
    "name" to event.name,
    "resourceId" to event.resourceId,
    "delivered" to event.delivered,
)

// -------------------------------------------------------------------- access rules

fun requireAdmin(user: User) {
    if (user.role != "admin") throw forbidden()
}

fun checkIfMatch(header: String?, version: Int) {
    if (header.isNullOrEmpty()) throw preconditionRequired()
    if (header != version.toString()) throw preconditionFailed()
}

fun checkIncludeDeleted(raw: String?, user: User): Boolean {
    if (raw == null) return false
    if (user.role != "admin") throw forbidden()
    return raw == "true"
}

// ---------------------------------------------------------------------- pagination

/** Sort by the tiebreak first, then stably by the requested field. */
fun paginate(rows: List<Map<String, Any?>>, page: Page): Map<String, Any?> {
    val tiebreak = if (rows.isNotEmpty() && "seq" in rows[0]) "seq" else "id"
    val byField = compareBy<Map<String, Any?>> { it[page.sort] as Comparable<*>? }
    val byTiebreak = compareBy<Map<String, Any?>> { it[tiebreak] as Comparable<*>? }
    val ordered = rows.sortedWith(
        (if (page.order == "desc") byField.reversed() else byField).then(byTiebreak)
    )
    return mapOf(
        "items" to ordered.drop(page.offset).take(page.limit),
        "total" to ordered.size,
        "limit" to page.limit,
        "offset" to page.offset,
    )
}

// --------------------------------------------------------------------------- tasks

fun readNote(
    actor: User,
    body: Map<String, Any?>,
    errors: MutableList<Map<String, String>>,
    current: String,
): String {
    if ("internalNote" !in body) return current
    if (actor.role != "admin") throw forbidden()
    val note = body["internalNote"] as? String ?: throw badRequest()
    if (note.length > MAX_TITLE_LENGTH) {
        errors.add(fail("internalNote", "internalNote is too long"))
    }
    return note
}

// ---------------------------------------------------------------- queries, reports

private fun hit(type: String, id: Int, label: String): Map<String, Any?> =
    mapOf("type" to type, "id" to id, "label" to label)

private fun group(key: String, picked: List<Task>): Map<String, Any?> = mapOf(
    "key" to key,
    "tasks" to picked.size,
    "totalScore" to picked.sumOf { computeScore(it.priority, it.status) },
)

fun checkBulkSize(operations: Any?) {
    if (operations !is List<*> || operations.isEmpty() || operations.size > MAX_BULK_ITEMS) {
        throw invalid(listOf(fail("operations", "operations is out of range")))
    }
}

class Service(private val store: Store) {

    // ----------------------------------------------------------------- serializers

    fun serializeProject(project: Project): Map<String, Any?> = mapOf(
        "id" to project.id,
        "name" to project.name,
        "ownerId" to project.ownerId,
        "taskCount" to store.taskCount(project.id),
        "version" to project.version,
        "deleted" to project.deleted,
    )

    // ---------------------------------------------------------------- access rules

    fun authenticate(header: String?): Caller {
        val token = if (header != null && header.startsWith("Bearer ")) header.substring(7) else ""
        val session = store.sessions[token] ?: throw unauthorized()
        val user = store.findUser(session.userId) ?: throw unauthorized()
        return Caller(user, session)
    }

    fun chargeQuota(user: User, session: Session): Int {
        if (session.used >= user.quota) throw quotaExceeded()
        val charged = session.copy(used = session.used + 1)
        store.sessions[charged.token] = charged
        return maxOf(user.quota - charged.used, 0)
    }

    fun reachableProject(projectId: Int?, user: User, includeDeleted: Boolean = false): Project {
        val project = store.findProject(projectId, includeDeleted) ?: throw notFound()
        if (user.role != "admin" && project.ownerId != user.id) throw forbidden()
        return project
    }

    fun reachableTask(taskId: Int?, user: User, includeDeleted: Boolean = false): Task {
        val task = store.findTask(taskId, includeDeleted) ?: throw notFound()
        reachableProject(task.projectId, user, true)
        return task
    }

    // ------------------------------------------------------------------------ auth

    fun login(username: String, password: String, token: String): User {
        val user = store.findByUsername(username)
        if (user == null || user.password != password) throw invalidCredentials()
        store.sessions[token] = Session(token, user.id)
        return user
    }

    // -------------------------------------------------------------------- projects

    fun createProject(actor: User, name: String, ownerId: Int?): Project {
        val errors = mutableListOf<Map<String, String>>()
        checkString(name, "name", MAX_NAME_LENGTH, errors)
        if (store.findUser(ownerId) == null) {
            errors.add(fail("ownerId", "ownerId is not a known user"))
        }
        if (errors.isNotEmpty()) throw invalid(errors)
        val taken = store.projects.values.any {
            it.ownerId == ownerId && it.name == name && !it.deleted
        }
        if (taken) throw conflict()
        val project = store.insertProject(name, ownerId!!)
        store.record(actor.id, "create", "project", project.id)
        return project
    }

    fun renameProject(actor: User, project: Project, name: String): Project {
        val errors = mutableListOf<Map<String, String>>()
        checkString(name, "name", MAX_NAME_LENGTH, errors)
        if (errors.isNotEmpty()) throw invalid(errors)
        val taken = store.projects.values.any {
            it.ownerId == project.ownerId && it.name == name && it.id != project.id && !it.deleted
        }
        if (taken) throw conflict()
        val renamed = project.copy(name = name, version = project.version + 1)
        store.projects[renamed.id] = renamed
        store.record(actor.id, "update", "project", renamed.id)
        return renamed
    }

    fun deleteProject(actor: User, project: Project): Project {
        val deleted = project.copy(version = project.version + 1, deleted = true)
        store.projects[deleted.id] = deleted
        store.record(actor.id, "delete", "project", deleted.id)
        for (task in store.liveTasksOf(project.id)) {
            deleteTask(actor, task)
        }
        return deleted
    }

    fun restoreProject(actor: User, project: Project): Project {
        if (!project.deleted) throw conflict()
        val restored = project.copy(version = project.version + 1, deleted = false)
        store.projects[restored.id] = restored
        store.record(actor.id, "restore", "project", restored.id)
        return restored
    }

    // ----------------------------------------------------------------------- tasks

    fun createTask(
        actor: User,
        project: Project,
        title: String,
        priority: Int?,
        assigneeId: Int?,
        note: String,
        errors: MutableList<Map<String, String>>,
    ): Task {
        checkString(title, "title", MAX_TITLE_LENGTH, errors)
        checkPriority(priority, errors)
        if (assigneeId != null && store.findUser(assigneeId) == null) {
            errors.add(fail("assigneeId", "assigneeId is not a known user"))
        }
        if (errors.isNotEmpty()) throw invalid(errors)
        val task = store.insertTask(project.id, title, priority!!, assigneeId, note)
        store.record(actor.id, "create", "task", task.id)
        return task
    }

    fun replaceTask(
        actor: User,
        task: Task,
        title: String,
        priority: Int?,
        assigneeId: Int?,
        note: String,
        errors: MutableList<Map<String, String>>,
    ): Task {
        checkString(title, "title", MAX_TITLE_LENGTH, errors)
        checkPriority(priority, errors)
        if (assigneeId != null && store.findUser(assigneeId) == null) {
            errors.add(fail("assigneeId", "assigneeId is not a known user"))
        }
        if (errors.isNotEmpty()) throw invalid(errors)
        val replaced = task.copy(
            title = title, priority = priority!!, assigneeId = assigneeId, internalNote = note,
            version = task.version + 1,
        )
        store.tasks[replaced.id] = replaced
        store.record(actor.id, "update", "task", replaced.id)
        return replaced
    }

    fun moveStatus(actor: User, task: Task, status: Any?): Task {
        val errors = mutableListOf<Map<String, String>>()
        checkStatus(status, errors)
        if (errors.isNotEmpty()) throw invalid(errors)
        val next = status as String
        if ((task.status to next) !in TRANSITIONS) throw invalidTransition()
        val moved = task.copy(status = next, version = task.version + 1)
        store.tasks[moved.id] = moved
        store.record(actor.id, "update", "task", moved.id)
        return moved
    }

    fun deleteTask(actor: User, task: Task): Task {
        val deleted = task.copy(version = task.version + 1, deleted = true)
        store.tasks[deleted.id] = deleted
        store.record(actor.id, "delete", "task", deleted.id)
        return deleted
    }

    fun restoreTask(actor: User, task: Task): Task {
        if (!task.deleted) throw conflict()
        val restored = task.copy(version = task.version + 1, deleted = false)
        store.tasks[restored.id] = restored
        store.record(actor.id, "restore", "task", restored.id)
        return restored
    }

    // -------------------------------------------------------------------- comments

    fun createComment(actor: User, task: Task, body: String): Comment {
        val errors = mutableListOf<Map<String, String>>()
        checkString(body, "body", MAX_COMMENT_LENGTH, errors)
        if (errors.isNotEmpty()) throw invalid(errors)
        val comment = store.insertComment(task.id, actor.id, body)
        store.record(actor.id, "create", "comment", comment.id)
        return comment
    }

    fun removeComment(actor: User, comment: Comment) {
        if (actor.role != "admin" && comment.authorId != actor.id) throw forbidden()
        store.comments.remove(comment.id)
        store.record(actor.id, "delete", "comment", comment.id)
    }

    // ----------------------------------------------------------------------- users

    fun createUser(actor: User, username: String, password: String, role: Any?, quota: Any?): User {
        val errors = mutableListOf<Map<String, String>>()
        checkString(username, "username", MAX_NAME_LENGTH, errors)
        checkString(password, "password", MAX_NAME_LENGTH, errors)
        checkRole(role, errors)
        checkQuota(quota, errors)
        if (errors.isNotEmpty()) throw invalid(errors)
        if (store.findByUsername(username) != null) throw conflict()
        val user = store.insertUser(username, password, role as String, quota as Int)
        store.record(actor.id, "create", "user", user.id)
        return user
    }

    fun updateUser(actor: User, user: User, body: Map<String, Any?>): User {
        val errors = mutableListOf<Map<String, String>>()
        if ("role" in body) checkRole(body["role"], errors)
        if ("quota" in body) checkQuota(body["quota"], errors)
        if (errors.isNotEmpty()) throw invalid(errors)
        val updated = user.copy(
            role = if ("role" in body) body["role"] as String else user.role,
            quota = if ("quota" in body) body["quota"] as Int else user.quota,
            version = user.version + 1,
        )
        store.users[updated.id] = updated
        store.record(actor.id, "update", "user", updated.id)
        return updated
    }

    fun deleteUser(actor: User, user: User): User {
        if (user.id == actor.id) throw conflict()
        val deleted = user.copy(version = user.version + 1, deleted = true)
        store.users[deleted.id] = deleted
        store.record(actor.id, "delete", "user", deleted.id)
        return deleted
    }

    // ------------------------------------------------------------ queries, reports

    fun visibleProjects(user: User, includeDeleted: Boolean): List<Project> =
        store.projects.values.filter {
            (includeDeleted || !it.deleted) && (user.role == "admin" || it.ownerId == user.id)
        }

    fun visibleTasks(user: User, includeDeleted: Boolean): List<Task> {
        val allowed = visibleProjects(user, true).map { it.id }.toSet()
        return store.tasks.values.filter {
            it.projectId in allowed && (includeDeleted || !it.deleted)
        }
    }

    fun search(user: User, query: String): Map<String, Any?> {
        val needle = query.lowercase()
        val results = visibleProjects(user, false)
            .filter { needle in it.name.lowercase() }
            .map { hit("project", it.id, it.name) } +
            visibleTasks(user, false)
                .filter { needle in it.title.lowercase() }
                .map { hit("task", it.id, it.title) }
        return mapOf("results" to results, "total" to results.size)
    }

    fun workload(user: User, groupBy: String): Map<String, Any?> {
        val rows = visibleTasks(user, false)
        val groups = when (groupBy) {
            "status" -> STATUSES.map { status ->
                group(status, rows.filter { it.status == status })
            }
            "assignee" -> {
                val named = rows.mapNotNull { it.assigneeId }.distinct().sorted()
                val loose = rows.filter { it.assigneeId == null }
                named.map { assignee ->
                    group(assignee.toString(), rows.filter { it.assigneeId == assignee })
                } + if (loose.isEmpty()) emptyList() else listOf(group("unassigned", loose))
            }
            else -> visibleProjects(user, false).map { project ->
                group(project.name, rows.filter { it.projectId == project.id })
            }
        }
        return mapOf("groupBy" to groupBy, "groups" to groups)
    }

    fun flushOutbox(): Int {
        val pending = store.outbox.withIndex().filter { !it.value.delivered }
        pending.forEach { (index, event) -> store.outbox[index] = event.copy(delivered = true) }
        return pending.size
    }

    fun metrics(): Map<String, Any?> = mapOf(
        "requests" to store.requests,
        "byStatus" to store.byStatus.entries.associate { (code, count) ->
            code.toString() to count
        },
        "byRoute" to store.byRoute.map { (route, count) ->
            mapOf("route" to route, "count" to count)
        },
        "auditEntries" to store.audit.size,
        "outboxPending" to store.outboxPending(),
    )

    fun stats(): Map<String, Any?> {
        val live = store.tasks.values.filter { !it.deleted }
        val byStatus = STATUSES.associateWith { status -> live.count { it.status == status } }
        val sumScore = live.sumOf { computeScore(it.priority, it.status) }
        val best = store.projects.values
            .filter { !it.deleted }
            .maxByOrNull { store.taskCount(it.id) }
        return mapOf(
            "projects" to store.projects.values.count { !it.deleted },
            "tasks" to live.size,
            "users" to store.users.values.count { !it.deleted },
            "sessions" to store.sessions.size,
            "comments" to store.comments.size,
            "byStatus" to byStatus,
            "avgScore" to if (live.isEmpty()) 0.0
            else Math.round(sumScore.toDouble() / live.size * 100) / 100.0,
            "topProjectName" to best?.name,
            "auditEntries" to store.audit.size,
            "outboxPending" to store.outboxPending(),
        )
    }
}
